chainparams: Change nChainTx type to uint64_t

fjahr opened the pull request:
This picks up the work from #29331 and closes #29258.

This simply changes the type and addresses the comments from #29331 by changing the type in all relevant places and removing unnecessary casts. This also adds an extremely simple unit test.

Additionally this modernizes the name of `nChainTx` which helps reviewers check all use of the symbol and can make silent merge conflicts.

fjahr committed:
chainparams: Change nChainTx to uint64_t

Also update types of assumeutxo chainparams and some related local variables for
consistency.

Co-authored-by: russeree <reese.russell@ymail.com>

fjahr committed:
test: Add basic check for nChainTx type

fanquake commented:
```bash
==25794==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 152 byte(s) in 1 object(s) allocated from:
    #0 0x5597c4ef3931 in operator new(unsigned long) (/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/test_bitcoin+0x1276931) (BuildId: dc550d0a0634dbe1f98f193b152915bbb7968f28)
    #1 0x5597c53b3398 in CreateBlockIndexWithNbits(unsigned int) src/test/blockchain_tests.cpp:24:32
    #2 0x5597c53b1e58 in blockchain_tests::num_chain_tx_max_invoker() src/test/blockchain_tests.cpp:77:1
    #3 0x5597c4fea063 in boost::function0<void>::operator()() const /usr/include/boost/function/function_template.hpp:771:14
    #4 0x5597c5070afa in boost::detail::forward::operator()() /usr/include/boost/test/impl/execution_monitor.ipp:1395:32
    #5 0x5597c5070afa in boost::detail::function::function_obj_invoker0<boost::detail::forward, int>::invoke(boost::detail::function::function_buffer&) /usr/include/boost/function/function_template.hpp:137:18
    #6 0x5597c4f2922f in int boost::detail::do_invoke<boost::shared_ptr<boost::detail::translator_holder_base>, boost::function<int ()>>(boost::shared_ptr<boost::detail::translator_holder_base> const&, boost::function<int ()> const&) /usr/include/boost/test/impl/execution_monitor.ipp:308:30
    #7 0x5597c4f2922f in boost::execution_monitor::catch_signals(boost::function<int ()> const&) /usr/include/boost/test/impl/execution_monitor.ipp:910:16
    #8 0x5597c4f297fe in boost::execution_monitor::execute(boost::function<int ()> const&) /usr/include/boost/test/impl/execution_monitor.ipp:1308:16
    #9 0x5597c4f214fb in boost::execution_monitor::vexecute(boost::function<void ()> const&) /usr/include/boost/test/impl/execution_monitor.ipp:1404:5
    #10 0x5597c4f214fb in boost::unit_test::unit_test_monitor_t::execute_and_translate(boost::function<void ()> const&, unsigned long) /usr/include/boost/test/impl/unit_test_monitor.ipp:49:9
    #11 0x5597c4fa3c92 in boost::unit_test::framework::state::execute_test_tree(unsigned long, unsigned long, boost::unit_test::framework::state::random_generator_helper const*) /usr/include/boost/test/impl/framework.ipp:815:44
    #12 0x5597c4fa2c0f in boost::unit_test::framework::state::execute_test_tree(unsigned long, unsigned long, boost::unit_test::framework::state::random_generator_helper const*) /usr/include/boost/test/impl/framework.ipp:784:58
    #13 0x5597c4fa2c0f in boost::unit_test::framework::state::execute_test_tree(unsigned long, unsigned long, boost::unit_test::framework::state::random_generator_helper const*) /usr/include/boost/test/impl/framework.ipp:784:58
    #14 0x5597c4f1f4a5 in boost::unit_test::framework::run(unsigned long, bool) /usr/include/boost/test/impl/framework.ipp:1722:29
    #15 0x5597c4f54a47 in boost::unit_test::unit_test_main(boost::unit_test::test_suite* (*)(int, char**), int, char**) /usr/include/boost/test/impl/unit_test_main.ipp:250:9
    #16 0x7f0c325b31c9  (/lib/x86_64-linux-gnu/libc.so.6+0x2a1c9) (BuildId: b58ab3be268281d0cf45f2b22cbd4ba8b4aac9bd)
    #17 0x7f0c325b328a in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2a28a) (BuildId: b58ab3be268281d0cf45f2b22cbd4ba8b4aac9bd)
    #18 0x5597c4e17804 in _start (/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/test_bitcoin+0x119a804) (BuildId: dc550d0a0634dbe1f98f193b152915bbb7968f28)

SUMMARY: AddressSanitizer: 152 byte(s) leaked in 1 allocation(s).
```

fjahr commented:
Fixed CI

Randy808 commented:
ACK afa67033772d750a15231aa0aa998a0454848bea

fjahr committed:
scripted-diff: Modernize naming of nChainTx and nTxCount

-BEGIN VERIFY SCRIPT-
sed -i 's/nChainTx/m_chain_tx_count/g' $(git grep -l 'nChainTx' ./src)
sed -i 's/nTxCount/tx_count/g' $(git grep -l 'nTxCount' ./src)
-END VERIFY SCRIPT-

fjahr commented:
Added a scripted diff renaming `nChainTx` to `m_chain_tx_count`.

fjahr commented on src/kernel/chainparams.h:
```diff
@@ -69,7 +69,7 @@ struct AssumeutxoData {
 struct ChainTxData {
     int64_t nTime;    //!< UNIX timestamp of last known number of transactions
-    int64_t nTxCount; //!< total number of transactions between genesis and that timestamp
+    uint64_t nTxCount; //!< total number of transactions between genesis and that timestamp
```
sure, added

fjahr commented:
@maflcko @ryanofsky curious about your thoughts on https://github.com/fjahr/bitcoin/commit/b9382041e541569a2b8ec7ed1f7fb10aa602277f , i.e. if should take that route with memory optimization or if you think it's not necessary after all.

fjahr commented:
> I don't think it is needed. If it was done, it should be a bitfield instead, no? See also [#29258 (comment)](https://github.com/bitcoin/bitcoin/issues/29258#issuecomment-1934619640) ?

I could change the approach to bitfields, sure. I don't think we are using them anywhere yet so I wasn't sure if it would really be the preferred approach. But if it's not needed, I will just leave the PR as is for now.

fjahr commented:
> Concept ACK. I think the pull description can be rewritten, since it ends up in plain text in the merge commit. Instead of writing "Thing A is not included. It is now", it would be better to say why it is included and remove the formatting/negation: "Thing A is included, because (...motivation...)"

I have updated the description.

glozow commented on src/chain.h:
```diff
@@ -174,7 +174,7 @@ class CBlockIndex
     //! to the genesis block or an assumeutxo snapshot block have reached the
     //! VALID_TRANSACTIONS level.
     //! Change to 64-bit type before 2024 (assuming worst case of 60 byte transactions).
```
64d35be20c43f764e5c22657085a790d0adcfe22 should remove this todo comment

fjahr replied:
Done

glozow reviewed (commented):
concept ACK

naiyoma commented on src/rpc/blockchain.cpp:
```diff
@@ -2716,7 +2716,7 @@ UniValue CreateUTXOSnapshot(
     result.pushKV("base_height", tip->nHeight);
     result.pushKV("path", path.utf8string());
     result.pushKV("txoutset_hash", maybe_stats->hashSerialized.ToString());
-    result.pushKV("nchaintx", tip->nChainTx);
```
Would it make sense to also change this string `nchaintx` to `m_chain_tx_count` for consistency?

fjahr replied:
This would mean that we return a different key in the JSON returned from the RPC. We usually don't do this unless it's needed in order to not break things on the user side. We also only use the `m_` prefix for class members.

naiyoma commented:
Concept ACK.
Nit: IMO,  [https://github.com/bitcoin/bitcoin/commit/64d35be20c43f764e5c22657085a790d0adcfe22](https://github.com/bitcoin/bitcoin/commit/64d35be20c43f764e5c22657085a790d0adcfe22) message should be amended to also include `nTxCount` and `nTxDiff`, since it's not just `nChainTx` that's changing to uint64_t.

fjahr commented:
Rebased and addressed comments

> Nit: IMO, https://github.com/bitcoin/bitcoin/commit/64d35be20c43f764e5c22657085a790d0adcfe22 message should be amended to also include nTxCount and nTxDiff, since it's not just nChainTx that's changing to uint64_t.

Naming all the local variables that are touched isn't necessary I think but I added a comment in the commit description.

fjahr commented:
Rebased (caused by #30329)

maflcko reviewed (commented):
ACK 242116cdf3fac6fa9af1d01387cca2cf4be1ccb 🕕

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: ACK 242116cdf3fac6fa9af1d01387cca2cf4be1ccb 🕕
KrtFcepn0TtAJ0W7GfxawMzY3jMtWFrv0xnB2Ns8BxeEtkJRw9IZoH2BkP3m719p6mBJGmYKIIPBIZZLjpECDw==
```

</details>

fjahr commented on src/test/blockchain_tests.cpp:
```diff
@@ -76,4 +76,11 @@ BOOST_AUTO_TEST_CASE(get_difficulty_for_very_high_target)
+{
+    std::unique_ptr<CBlockIndex> block_index{CreateBlockIndexWithNbits(0x1f111111)};
+    block_index->nChainTx = std::numeric_limits<uint64_t>::max();
+    BOOST_CHECK_EQUAL(block_index->nChainTx, std::numeric_limits<uint64_t>::max());
```
> nit in [e78cf83](https://github.com/bitcoin/bitcoin/commit/e78cf83beec2f508823803947bdc3d9bcb7c7217): Not sure why nbits matter for this test.

Yeah, I had just reused the helper method from this test but it's not needed indeed. I am using an empty constructor now.

> I think you can just write `CBlockIndex block{.nChainTx=..max()};`, replacing the first two lines.

I am unfamiliar with the `..max()` syntax and I am also getting a compiler error when I try to use a designated initializer here, I didn't investigate deeper why that is though.

fjahr commented:
> I think you can drop the new name from the pull description. It is wrong right now (doesn't match the scripted diff), and if anyone wanted to find it, they can look at the scripted diff.

done

fjahr commented:
Rebased

marcofleon reviewed (commented):
ACK fab49990eb787aabd908f6ba88f5e00ed9844f91. Reviewed the changes, grepped for the old variable names, and ran `make check` and the individual test. I read a bit of the background (links provided in the PR description) and there doesn't seem to be any issue with this change in `validation`.

glozow reviewed (commented):
utACK fab49990eb787aabd908f6ba88f5e00ed9844f91, happy to reack after rebase

BrandonOdiwuor reviewed (commented):
Code Review ACK fab49990eb787aabd908f6ba88f5e00ed9844f91

Looks Good To Me

fjahr commented:
Thanks for the reviews! Rebased now.

glozow commented:
reACK bf0efb4fc72 via range-diff
